[AI-4493] DDS: Cisco Secure Web Appliance Integration v1.0.0 (#18717)
DataDog/integrations-core: cisco_secure_web_appliance/README.md

## Overview

[Cisco Secure Web Appliance][4] protects your organization by automatically blocking risky sites and testing unknown sites before allowing users access. It intercepts and monitors Internet traffic and applies policies to help keep your internal network secure from malware, sensitive data loss, productivity loss, and other Internet-based threats.

This integration ingests the following types of logs:
- Access Logs: This records all Web Proxy filtering and scanning activity.
- L4TM Logs: This records all Layer-4 Traffic Monitor activity.

> Visualize detailed insights into Web Proxy filtering and scanning activity and Layer-4 Traffic Monitor activity through the out-of-the-box dashboards.Additionally, out-of-the-box detection rules are available to help you monitor and respond to potential security threats effectively.

Author: Updated.

1. Collecting logs is disabled by default in the Datadog Agent. Enable it in datadog.yaml:

    ```yaml
    logs_enabled: true
    ```
2. If the chosen retrieval method for Access Logs is Syslog Push, add the configuration block for Access Logs in the configuration file to start collecting your Cisco Secure Web Appliance Access Logs along with L4TM Logs. The configuration will appear as follows in cisco_secure_web_appliance.d/conf.yaml.

    We will be using the UDP method to collect the Cisco Secure Web Appliance Access Logs.
    See the sample [cisco_secure_web_appliance.d/conf.yaml][7] for available configuration options.

    ```yaml
      logs:
      - type: file
        path: <Path to Directory Where L4TM Logs would Get Stored>
        service: l4tm_logs
        source: cisco-secure-web-appliance
      logs:
      - type: udp
        port: <PORT>
        service: access_logs
        source: cisco-secure-web-appliance
    ```
    **Note**: It is important not to change the service and source values, as these parameters are integral to the pipeline's operation.
    
3. [Restart the Agent][3].

#### Steps to configure Syslog Push for Access Logs:

**Prerequisites:**

> 1. Require Syslog server hostname(Your Datadog Agent Server) where you want to push the logs.

Reviewer: ```suggestion
- The Syslog server hostname (your Datadog Agent Server) where you want to push the logs.
```

Author: Updated.

**Configuration:**

> 1. Login to Cisco Secure Web Appliance UI.

Reviewer: ```suggestion
1. Log in to Cisco Secure Web Appliance UI.
```

Author: Updated.

2. Navigate to  System Administration -> Log Subscriptions.

> 3. In order to add  Access Logs subscription, click on the Add `Log Subscription` button.

Reviewer: ```suggestion
3. In order to add  Access Logs subscription, click on the **Add Log Subscription** button.
```

Author: Updated.

4. Select Log Type as Access Logs.
5. Provide Log Name.

> 6. Choose `Squid` Option for Log Style.\

Reviewer: ```suggestion
6. Choose the `Squid` option for Log Style.
```

Author: Updated.

>     **Note**: We will support default (squid) log style for access log. 

Reviewer: ```suggestion
    **Note**: The default (squid) log style for access logs is supported. 
```

Author: Updated as per the suggestion.

> 7. Select `Syslog Push` Option as Retrieval Method.

Reviewer: ```suggestion
7. Select `Syslog Push` option as the Retrieval Method.
```

Author: Updated as per the suggestion.

8. Provide Following Details.

    Hostname: \<Datadog-Agent Host Server>

    Port: \<Default Provided>

    Protocol: UDP
    
    Maximum message size: \<Valid values for UDP are 1024 to 9216>
    
    Facility: \<Default Selected>
9. Click on Submit. 

> 10. Click on Commit Changes at top right-side of Log Subscriptions Page.
>     NOTE: These changes will not go into effect until you commit them.

Reviewer: ```suggestion
10. Click on **Commit Changes** at the top right of the Log Subscriptions page.
    **Note:** These changes will not go into effect until they are committed.
```

Author: Updated as per the suggestion.

> #### Steps to configure SCP on Remote Server for L4TM Logs.

Reviewer: ```suggestion
#### Configure SCP on Remote Server for L4TM Logs
```

Author: Updated as per the suggestion.

**Configuration:**
1. Navigate to  System Administration -> Log Subscriptions in Cisco Secure Web Appliance UI.
2. To add a log subscription for Traffic Monitor Logs, click on  Add `Log Subscription`.  
3. Select Traffic Monitor Logs as Log Type.
4. Provide appropriate Log Name.
5. For FileName, provide a new name or keep the default added name.
6. Choose SCP on Remote Server as Retrieval Method.
7. Provide Following Information.

    SCP Host: \<SCP Host IP Address>

    Directory: \<Path to Directory Where Logs would Get Stored>
    
    **NOTE:** Make sure that Directory does not have any other log files.
    SCP Port: \<Default Port>
    Username: \<SCP Host Username>
8. Click on Submit. After submitting, SSH key(s)  will get generated. Copy and save both the SSH key(s) as it is only visible once.
9. Place the given SSH key(s) into your `authorized_keys` file on the remote host so that the log files can be uploaded.
10. Click on Commit Changes at top right-side of Log Subscriptions Page.

    **NOTE:** These changes will not go into effect until you commit them.

#### Steps to configure SCP on Remote Server for Access Logs.

**Prerequisites:**

> 1. Requires SCP Host and Username for the same.

Reviewer: Not sure what this means.
Does it mean the you need the SCP Host and its admin account username?

Author: Understood your confusion. this basically means that it requires hostname and username (admin account username is not necessary) of VM/machine on which datadog-agent is installed. Also we have updated the sentence accordingly.

**Configuration:**

> 1. Go to System Administration -> Log Subscriptions in Cisco Secure Web Appliance UI.
> 2. To add a new log subscription for Access Logs, click Add `Log Subscription`  OR edit already configured Access Logs Subscription.
> 3. If you are adding a new subscription then follow the steps 4 to 6 mentioned in the configuration of Syslog Push for Access Logs.
> 4. If you are editing existing Access Logs Subscription, select SCP on Remote Server as Retrieval Method.
> 5. Provide following information:
> 
>     SCP Host: \<SCP Hostname>
> 
>     SCP Port: \<Default Provided>
>     
>     Directory: \<Path to store the Log Files>
>     
>     **NOTE:** Make sure that Directory does not have any other log files.
> 10. Username: \<SCP Server Username>
> 11. Click on Submit. Once you click on `Submit`,  SSH key(s) will get generated. Copy the SSH Key and save it somewhere as this will be only displayed once. 
> 12. Place the given SSH key(s) into your `authorized_keys` file on the remote host so that the log files can be uploaded.
> Click on Commit Changes at top right-side of Log Subscriptions Page.
> 
>     **NOTE:** These changes will not go into effect until you commit them.

Reviewer: ```suggestion
1. In the Cisco Secure Web Appliance UI, go to System Administration, and then Log Subscriptions.
2. To add a new log subscription for Access Logs, click **Add Log Subscription** or edit an existing Access Logs Subscription.
3. If you are adding a new subscription, then follow steps 4 to 6 mentioned in the Syslog Push for Access Logs section.
4. If you are editing an existing Access Logs Subscription, select SCP on the Remote Server as the Retrieval Method.
5. Provide following information:

    SCP Host: \<SCP Hostname>

    SCP Port: \<Default Provided>
    
    Directory: \<Path to store the Log Files>
    
    **Note:** Make sure that Directory does not have any other log files.
10. Username: \<SCP Server Username>
11. Click **Submit**. Once you click **Submit**,  SSH key(s) are generated. Copy the SSH Key and save it somewhere as this will be only displayed once. 
12. Place the given SSH key(s) into your `authorized_keys` file on the remote host so that the log files can be uploaded.
13. Click **Commit Changes** at top right of Log Subscriptions Page.

    **Note:** These changes do not go into effect until you commit them.
```

Author: Updated as per the suggestion.

#### Steps to set time zone to GMT in Cisco Secure Web Appliance 

> Since Datadog expects all the logs in GMT timezone by default, if the time zone of your Cisco Secure Web Appliance logs is other than GMT, please change it to GMT. Here are the steps:
> 1. Go to System Administration->Time Zone.
> 2. Click on `Edit Settings`.
> 3. Select GMT Offset as the region.
> 4. Select GMT as the Country.
> 5. Select GMT (GMT) as the time zone.
> 6. Submit and commit the changes.

Reviewer: ```suggestion
Datadog expects all the logs in GMT time zone by default. If the time zone of your Cisco Secure Web Appliance logs is not GMT, please change it to GMT. Here are the steps:

1. Go to **System Administration**, and then **Time Zone**.
2. Click on **Edit Settings**.
3. Select **GMT Offset** as the region.
4. Select **GMT** as the country.
5. Select **GMT (GMT)** as the time zone.
6. Submit and commit the changes.
```

If you see a **Permission denied** error while port binding in the Agent logs, see the following instructions:

1. Binding to a port number under 1024 requires elevated permissions. Grant access to the port using the `setcap` command:
    ```shell
    sudo setcap CAP_NET_BIND_SERVICE=+ep /opt/datadog-agent/bin/agent/agent
    ```

2. Verify the setup is correct by running the `getcap` command:

    ```shell
    sudo getcap /opt/datadog-agent/bin/agent/agent
    ```

    With the expected output:

    ```shell
    /opt/datadog-agent/bin/agent/agent = cap_net_bind_service+ep
    ```

    **Note**: Re-run this `setcap` command every time you upgrade the Agent.
3. [Restart the Agent][3].

**Permission denied while file monitoring:**

If you see a **Permission denied** error while monitoring the log files, give the `dd-agent` user read permission on them.
  ```shell
    sudo chmod g+s Path/to/Directory/Where/Logs/would/Get/Stored/
  ```
  ```shell
    sudo chgrp dd-agent Path/to/Directory/Where/Logs/would/Get/Stored/
  ```

**Data is not being collected:**

Make sure that traffic is bypassed from the configured port if the firewall is enabled.

**Port already in use:**

> - If you see the **Port \<PORT-NO> Already in Use** error, see the following instructions. The example below is for PORT-NO = 514:
> 
> - On systems using Syslog, if the Agent listens for access_logs logs on port 514, the following error can appear in the Agent logs: `Can't start UDP forwarder on port 514: listen udp :514: bind: address already in use`.
> 
> - This error occurs because by default, Syslog listens on port 514. To resolve this error, take **one** of the following steps: 

Reviewer: ```suggestion
If you see the **Port \<PORT-NO> Already in Use** error, see the following instructions. The example below is for PORT-NO = 514:

On systems using Syslog, if the Agent listens for access_logs logs on port 514, the following error can appear in the Agent logs: `Can't start UDP forwarder on port 514: listen udp :514: bind: address already in use`.

This error occurs because by default, Syslog listens on port 514. To resolve this error, take **one** of the following steps: 
```

Author: Updated as per the suggestion.

## Overview

[Cisco Secure Web Appliance][4] protects your organization by automatically blocking risky sites and testing unknown sites before allowing users access. It intercepts and monitors Internet traffic and applies policies to help keep your internal network secure from malware, sensitive data loss, productivity loss, and other Internet-based threats.

This integration ingests the following types of logs:
- Access Logs: This records all Web Proxy filtering and scanning activity.
- L4TM Logs: This records all Layer-4 Traffic Monitor activity.

Out-of-the-box dashboards help you to visualize detailed insights into Web Proxy filtering and scanning activity and Layer-4 Traffic Monitor activity. Additionally, out-of-the-box detection rules are available to help you monitor and respond to potential security threats effectively.

> **Disclaimer**: Your use of this integration, which may collect data that includes personal information, is subject to your agreements with Datadog. Cisco is not responsible for the privacy, security or integrity of any end-user information, including personal data, transmitted through your use of the integration.

Author: Updated as suggested

#### Logs Collection
**Tail File(File Monitoring)**

1. Collecting logs is disabled by default in the Datadog Agent. Enable it in datadog.yaml:

    ```yaml
    logs_enabled: true
    ```

2. Add this configuration block to your cisco_secure_web_appliance.d/conf.yaml file to start collecting your Cisco Secure Web Appliance L4TM Logs.

    ```yaml
      logs:
      - type: file
        path: <Path to Directory Where Logs would Get Stored>
        service: l4tm_logs
        source: cisco-secure-web-appliance
    ```

3. If the chosen retrieval method for Access Logs is SCP on Remote Server, add the configuration block for Access Logs in the above configuration to start collecting your Cisco Secure Web Appliance Access Logs along with L4TM Logs. The configuration will appear as follows in cisco_secure_web_appliance.d/conf.yaml.
    
    ```yaml
      logs:
      - type: file
        path: <Path to Directory Where L4TM Logs would Get Stored>
        service: l4tm_logs
        source: cisco-secure-web-appliance
      - type: file
        path: <Path to Directory Where Access Logs would Get Stored>
        service: access_logs
        source: cisco-secure-web-appliance
    ```

>     **NOTE**: Please make sure that `path` value should be similar to the Directory configured in `Configure SCP on Remote Server for L4TM Logs` and `Configure SCP on Remote Server for Access Logs` section respectively, forwarding /*.s

Author: Yes, files with the .s extension are generated and forwarded from the CSWA portal. To monitor all these files, we used /*.s, which tracks all files ending with the .s extension.

      logs:
      - type: file
        path: <Path to Directory Where L4TM Logs would Get Stored>
        service: l4tm_logs
        source: cisco-secure-web-appliance
      logs:
      - type: udp
        port: <PORT>
        service: access_logs
        source: cisco-secure-web-appliance
    ```
    **Note**: It is important not to change the service and source values, as these parameters are integral to the pipeline's operation.
    
3. [Restart the Agent][3].

### Configuration on Cisco Secure Web Appliance portal

#### Steps to set time zone to GMT
Datadog expects all the logs in GMT time zone by default. Please make sure that timezone configured in your Cisco Secure Web Appliance portal must be in GMT. Here are the steps to change the timzone:
1. Go to **System Administration**, and then **Time Zone**.
2. Click on **Edit Settings**.
3. Select **GMT Offset** as the region.
4. Select **GMT** as the country.
5. Select **GMT (GMT)** as the time zone.
6. Submit and commit the changes.

#### Configure Log Subscriptions

#### Configure Syslog Push for Access Logs:

**Prerequisites:**
- The datadog-agent server hostname where you want to push the logs.

**Configuration:**

> 1. Log in to Cisco Secure Web Appliance UI.
> 2. Navigate to  **System Administration -> Log Subscriptions**.
> 3. In order to add  Access Logs subscription, click on the **Add Log Subscription** button.
> 4. Select Log Type as Access Logs.
> 5. Provide Log Name.
> 6. Choose the `Squid` option for Log Style.
>     **Note**: The default (squid) log style for access logs is supported. 
> 7. Select `Syslog Push` option as the Retrieval Method.
> 8. Provide Following Details.

Reviewer: ```suggestion
1. Log in to the Cisco Secure Web Appliance UI.
2. Navigate to  **System Administration** > **Log Subscriptions**.
3. In order to add an Access Logs subscription, click **Add Log Subscription**.
4. Select **Log Type** as **Access Logs**.
5. Provide a Log Name.
6. Choose the **Squid** option for **Log Style**.
    **Note**: The default (squid) log style for access logs is supported. 
7. Select the **Syslog Push** option as the **Retrieval Method**.
8. Provide the following details.
```

Author: Updated as suggested

    Hostname: \<Datadog-Agent Host Server>

    Port: \<Default Provided>

    Protocol: UDP
    
    Maximum message size: \<Valid values for UDP are 1024 to 9216>
    
    Facility: \<Default Selected>

> 9. Click on Submit. 
> 10. Click on **Commit Changes** at the top right of the Log Subscriptions page.

Reviewer: ```suggestion
9. Click **Submit**. 
10. Click **Commit Changes** at the top right of the **Log Subscriptions** page.
```

Author: Updated as suggested

    **Note:** These changes will not go into effect until they are committed.

#### Configure SCP on Remote Server for L4TM Logs

**Prerequisites:**

> 1. Requires hostname and username (admin account username is not necessary) of VM/machine on which datadog-agent is installed.

Reviewer: ```suggestion
- Requires the hostname and username (admin account username is not necessary) of VM/machine where the Datadog Agent is installed.
```

Author: Updated as suggested

**Configuration:**

> 1. Navigate to  **System Administration -> Log Subscriptions** in Cisco Secure Web Appliance UI.
> 2. To add a log subscription for Traffic Monitor Logs, click on  Add **Log Subscription**.  
> 3. Select Traffic Monitor Logs as Log Type.
> 4. Provide appropriate Log Name.
> 5. For FileName, provide a new name or keep the default added name.
> 6. Choose SCP on Remote Server as Retrieval Method.
> 7. Provide Following Information.

Reviewer: ```suggestion
1. Navigate to  **System Administration** > **Log Subscriptions** in the Cisco Secure Web Appliance UI.
2. To add a log subscription for Traffic Monitor Logs, click **Add Log Subscription**.  
3. Select **Traffic Monitor Logs** as **Log Type**.
4. Provide the appropriate log name.
5. For **FileName**, provide a new name or keep the default name.
6. Choose **SCP on Remote Server** as **Retrieval Method**.
7. Provide the following information.
```

    SCP Host: \<SCP Host IP Address>

    Directory: \<Path to Directory Where Logs would Get Stored>
    **NOTE:** Make sure that Directory does not have any other log files.

    SCP Port: \<Default Port>

    Username: \<SCP Host Username>

> 8. Click on Submit. After submitting, SSH key(s)  will get generated. Copy and save both the SSH key(s) as it is only visible once.
> 9. Place the given SSH key(s) into your `authorized_keys` file on the remote host so that the log files can be uploaded.
> 10. Click on Commit Changes at top right-side of Log Subscriptions Page.

Reviewer: ```suggestion
8. Click **Submit**. After submitting, SSH key(s) are generated. Copy and save the SSH key(s) as it is only visible once.
9. Place the SSH key(s) into your `authorized_keys` file on the remote host so that the log files can be uploaded.
10. Click **Commit Changes** at top right of **Log Subscriptions** page.
```

    **NOTE:** These changes will not go into effect until you commit them.

#### Configure SCP on Remote Server for Access Logs.

**Prerequisites:**

> 1. Requires hostname and username (admin account username is not necessary) of VM/machine on which datadog-agent is installed.

Reviewer: ```suggestion
- Requires the hostname and username (admin account username is not necessary) of VM/machine where the Datadog Agent is installed.
```

Author: Updated as suggested

> 1. Requires hostname and username (admin account username is not necessary) of VM/machine on which datadog-agent is installed.
> 
> **Configuration:**
> 1. In the Cisco Secure Web Appliance UI, go to System Administration, and then Log Subscriptions.
> 2. To add a new log subscription for Access Logs, click **Add Log Subscription** or edit an existing Access Logs Subscription.
> 3. If you are adding a new subscription, then follow steps 4 to 6 mentioned in the `Configure Syslog Push for Access Logs` section.
> 4. If you are editing an existing Access Logs Subscription, select SCP on the Remote Server as the Retrieval Method.
> 5. Provide following information:
> 
>     SCP Host: \<SCP Hostname>
> 
>     SCP Port: \<Default Provided>
>     
>     Directory: \<Path to store the Log Files>
>     **Note:** Make sure that Directory does not have any other log files.
>     
>     Username: \<SCP Server Username>
> 6. Click **Submit**. Once you click **Submit**,  SSH key(s) are generated. Copy the SSH Key and save it somewhere as this will be only displayed once. 
> 7. Place the given SSH key(s) into your `authorized_keys` file on the remote host so that the log files can be uploaded.
> 8. Click **Commit Changes** at top right of Log Subscriptions Page.
>     **Note:** These changes do not go into effect until you commit them.

Reviewer: ```suggestion
- Requires the hostname and username (admin account username is not necessary) of VM/machine where the Datadog Agent is installed.

**Configuration:**
1. In the Cisco Secure Web Appliance UI, go to **System Administration** > **Log Subscriptions**.
2. To add a new log subscription for Access Logs, click **Add Log Subscription** or edit an existing Access Logs Subscription.
3. If you are adding a new subscription, then follow steps 4 to 6 mentioned in the Configure Syslog Push for Access Logs section or this topic.
4. If you are editing an existing Access Logs Subscription, select **SCP on the Remote Server** as the **Retrieval Method**.
5. Provide the following information:

    SCP Host: \<SCP Hostname>

    SCP Port: \<Default Provided>
    
    Directory: \<Path to store the Log Files>
    **Note:** Make sure that Directory does not have any other log files.
    
    Username: \<SCP Server Username>
6. Click **Submit**. Once you click **Submit**,  SSH key(s) are generated. Copy the SSH Key and save it somewhere as it is only displayed once. 
7. Place the SSH key(s) into your `authorized_keys` file on the remote host so log files can be uploaded.
8. Click **Commit Changes** at top right of the **Log Subscriptions** page.
    **Note:** These changes do not go into effect until you commit them.
```

> ##### For more information, Visit this [Cisco Secure Web Appliance official documentation for configuration][8]

Reviewer: ```suggestion
For more information on configuration, visit the [Cisco Secure Web Appliance official documentation][8].
```

Author: Updated as suggested

## Troubleshooting

**Permission denied while port binding:**

If you see a **Permission denied** error while port binding in the Agent logs, see the following instructions:

1. Binding to a port number under 1024 requires elevated permissions. Grant access to the port using the `setcap` command:
    ```shell
    sudo setcap CAP_NET_BIND_SERVICE=+ep /opt/datadog-agent/bin/agent/agent
    ```

2. Verify the setup is correct by running the `getcap` command:

    ```shell
    sudo getcap /opt/datadog-agent/bin/agent/agent
    ```

    With the expected output:

    ```shell
    /opt/datadog-agent/bin/agent/agent = cap_net_bind_service+ep
    ```

    **Note**: Re-run this `setcap` command every time you upgrade the Agent.
3. [Restart the Agent][3].

**Permission denied while file monitoring:**

If you see a **Permission denied** error while monitoring the log files, give the `dd-agent` user read permission on them.
  ```shell
    sudo chmod g+s Path/to/Directory/Where/Logs/would/Get/Stored/
  ```
  ```shell
    sudo chgrp dd-agent Path/to/Directory/Where/Logs/would/Get/Stored/
  ```

**Data is not being collected:**

> Make sure that traffic is bypassed from the configured port if the firewall is enabled.

Reviewer: ```suggestion
Ensure that traffic is bypassed from the configured port if the firewall is enabled.
```

Author: Updated as suggested
